### DISCUSSION TITLE
Migration from floating to reserved IPs

### DISCUSSION THREAD

**Discussion by bheisig** [digitalocean/terraform-provider-digitalocean#844] [Q&A] [answered]
Hello,

I've updated to the latest version of this provider. Is there a safe way to migrate from former floating IP resources to reserved ones without losing the assigned IPv4 addresses? I've tested it by just renaming them in all `.tf` files but ended in a broken Terraform state. I had to re-create reserved IPs and assign them to existing droplets. My previous assigned floating IPs are gone. Of course, I know of `terraform import`, `terraform state rm` and so on but failed anyway. Therefore, I'm interested in a step-by-step migration.

Thanks in advance.
Ben

**Comment by andrewsomething** [comment, marked as answer]
If you are using v2.21.0 of the provider, the `digitalocean_floating_ip` resource has been updated to use the `api.digitalocean.com/v2/reserved_ips` endpoints under the hood. So no immediate action is needed. That said, we will likely deprecate it at some point in the future to reduce code duplication and prevent new users from mistakenly choosing the wrong resource.

Starting from a config like: 

```hcl
resource "digitalocean_floating_ip" "foobar" {
  region = "nyc1"
}

resource "digitalocean_droplet" "foobar" {
  name               = "foobar"
  size               = "s-1vcpu-1gb"
  image              = "ubuntu-20-04-x64"
  region             = "nyc1"
}

resource "digitalocean_floating_ip_assignment" "foobar" {
  ip_address = digitalocean_floating_ip.foobar.ip_address
  droplet_id = digitalocean_droplet.foobar.id
}
``` 

1. First, import the reserved IP using the same IP as the existing floating IP (`terraform import digitalocean_reserved_ip.foobar x.x.x.x`) while adding the `digitalocean_reserved_ip` block to the config 
2. Next, change the `digitalocean_floating_ip_assignment` to a `digitalocean_reserved_ip_assignment` in the config
3. Then, remove the floating IP and `digitalocean_floating_ip_assignment` from state (`terraform state rm digitalocean_floating_ip.foobar && terraform state rm digitalocean_floating_ip_assignment.foobar`)
4. Finally, run `terraform apply` so the assignment info is updated in state. 

There are probably a few different ways to accomplish this. The most important part is removing the floating IP from the state before running an apply in order to retain the existing IP address. You don't want to delete the floating IP as it is the same as deleting the reserved IP.
